File: backend/src/ontoforge_server/runtime/embedding.py
from __future__ import annotations

import logging
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ontoforge_server.runtime.service import PropertyDef

logger = logging.getLogger(__name__)

# nomic-embed-text has 8192 token limit; ~4 chars/token → 30000 chars as safe threshold
_MAX_TEXT_CHARS = 30000


def build_text_repr(
    entity_type_key: str,
    properties: dict,
    property_defs: dict[str, PropertyDef],
) -> str:
    """Build a text representation of an entity for embedding.

    Format: "{entity_type_key}: {key}={value}, {key}={value}, ..."
    Only includes string properties with non-null values, in schema-defined order.
    """
    parts: list[str] = []
    for prop_key, prop_def in property_defs.items():
        if prop_def.data_type != "string":
            continue
        value = properties.get(prop_key)
        if value is not None:
            parts.append(f"{prop_key}={value}")

    text = f"{entity_type_key}: {', '.join(parts)}" if parts else entity_type_key

    if len(text) > _MAX_TEXT_CHARS:
        logger.warning(
            "Text representation for entity type '%s' truncated from %d to %d chars",
            entity_type_key,
            len(text),
            _MAX_TEXT_CHARS,
        )
        text = text[:_MAX_TEXT_CHARS]

    return text

File: server/src/runtime/embedding.ts
/**
 * Entity embedding text composition. Ports the Python reference
 * (`backend/src/ontoforge_server/runtime/embedding.py`) exactly — the
 * composition rules are normative in
 * `docs/capabilities/search.md#what-gets-embedded`:
 *
 * - only `string` properties contribute (never document/numeric/temporal);
 * - `key=value` pairs in schema declaration order;
 * - properties with no value are skipped;
 * - composed from the FULL schema, never the lens;
 * - capped at 30 000 characters;
 * - deterministic.
 */

import { cpLength, cpSlice } from "./codePoints.js";
import type { PropertyDef } from "./schemaCache.js";

// nomic-embed-text has an 8192-token limit; ~4 chars/token → 30000 chars
// as a safe threshold.
export const MAX_TEXT_CHARS = 30_000;

/**
 * Build the text an entity is embedded from:
 * `"{entityTypeKey}: {key}={value}, ..."` — or the bare type key when no
 * string property has a value.
 */
export function buildTextRepr(
  entityTypeKey: string,
  properties: Record<string, unknown>,
  propertyDefs: Record<string, PropertyDef>,
): string {
  const parts: string[] = [];
  for (const [propKey, propDef] of Object.entries(propertyDefs)) {
    if (propDef.dataType !== "string") {
      continue;
    }
    const value = properties[propKey];
    if (value !== null && value !== undefined) {
      parts.push(`${propKey}=${String(value)}`);
    }
  }

  let text = parts.length > 0 ? `${entityTypeKey}: ${parts.join(", ")}` : entityTypeKey;

  const length = cpLength(text);
  if (length > MAX_TEXT_CHARS) {
    console.warn(
      `Text representation for entity type '${entityTypeKey}' truncated ` +
        `from ${length} to ${MAX_TEXT_CHARS} chars`,
    );
    text = cpSlice(text, 0, MAX_TEXT_CHARS);
  }

  return text;
}
